sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnosticsContext.java

```diff
@@ -556,10 +581,233 @@ public String toJson() {
+     * included. Calling this method will lazily collect the user agent - which can be useful when writing
+     * a custom {@link CosmosDiagnosticsHandler}
+     * @return the UserAgent header value used for the client that issued this operation
+     */
+    public String getUserAgent() {
+        return this.userAgent;
+    }
+
+    /**
+     * Returns the connection mode used in the client.
+     * NOTE: this information is not included in the json representation returned from {@link #toJson()} because it
+     * is usually only relevant when thresholds are violated, in which case the entire diagnostics json-string is
+     * included. Calling this method will lazily collect the user agent - which can be useful when writing
+     * a custom {@link CosmosDiagnosticsHandler}
+     * @return the connection mode used in the client.
+     */
+    public String getConnectionMode() {
+        return this.connectionMode;
+    }
+
+    private static void addRequestInfoForGatewayStatistics(
+        ClientSideRequestStatistics requestStats,
+        ArrayList<CosmosDiagnosticsRequestInfo> requestInfo,
+        ClientSideRequestStatistics.GatewayStatistics gatewayStats) {
+
+        if (gatewayStats == null) {
+            return;
+        }
+
+        CosmosDiagnosticsRequestInfo info = new CosmosDiagnosticsRequestInfo(
+            requestStats.getActivityId(),
+            null,
+            gatewayStats.getPartitionKeyRangeId(),
+            gatewayStats.getResourceType() + ":" + gatewayStats.getOperationType(),
+            requestStats.getRequestStartTimeUTC(),
+            requestStats.getDuration(),
+            null,
+            gatewayStats.getRequestCharge(),
+            gatewayStats.getResponsePayloadSizeInBytes(),
+            gatewayStats.getStatusCode(),
+            gatewayStats.getSubStatusCode(),
+            new ArrayList<>()
+        );
+
+        requestInfo.add(info);
+    }
+
+    private static void addRequestInfoForStoreResponses(
+        ClientSideRequestStatistics requestStats,
+        ArrayList<CosmosDiagnosticsRequestInfo> requestInfo,
+        List<ClientSideRequestStatistics.StoreResponseStatistics> storeResponses) {
+
+        for (ClientSideRequestStatistics.StoreResponseStatistics responseStats: storeResponses) {
+
+            StoreResultDiagnostics resultDiagnostics = responseStats.getStoreResult();
+            if (resultDiagnostics == null) {
+                continue;
+            }
+
+            StoreResponseDiagnostics responseDiagnostics = resultDiagnostics.getStoreResponseDiagnostics();
+
+            String partitionId = null;
+            String[] partitionAndReplicaId = resultDiagnostics.getPartitionAndReplicaId();
+            if (partitionAndReplicaId.length == 2) {
+                partitionId = partitionAndReplicaId[0];
+            }
+
+            Collection<CosmosDiagnosticsRequestEvent> events = new ArrayList<>();
```

Comment:
If ordering / sequencing is required, it can be a `List` correct?

Reply:
Fixed in next iteration